Source/WebCore/rendering/BaselineAlignment.h

```diff
@@ -0,0 +1,126 @@
+    // 'baseline-preference'
+    bool isCompatible(WritingMode, ItemPosition) const;
+
+    // Determines whether the baseline-sharing group's associated block-flow is opposite (LR vs RL) to particular
+    // item's writing-mode.
+    bool isOppositeBlockFlow(WritingMode blockFlow) const;
+
+    // Determines whether the baseline-sharing group's associated block-flow is orthogonal (vertical vs horizontal)
+    // to particular item's writing-mode.
+    bool isOrthogonalBlockFlow(WritingMode blockFlow) const;
+
+    WritingMode m_blockFlow;
+    ItemPosition m_preference;
+    LayoutUnit m_maxAscent;
+    HashSet<const RenderBox*> m_items;
+};
+
+// https://drafts.csswg.org/css-align-3/#shared-alignment-context
+// Boxes share an alignment context along a particular axis when they are:
+//
+//  * table cells in the same row, along the table's row (inline) axis
+//  * table cells in the same column, along the table's column (block) axis
+//  * grid items in the same row, along the grid's row (inline) axis
+//  * grid items in the same column, along the grid's colum (block) axis
+//  * flex items in the same flex line, along the flex container's main axis
+//
+// https://drafts.csswg.org/css-align-3/#baseline-sharing-group
+// A Baseline alignment-context may handle several baseline-sharing groups. In order to create an instance, we
+// need to pass the required data to define the first baseline-sharing group; a Baseline Context must have at
+// least one baseline-sharing group.
+//
+// By adding new items to a Baseline Context, the baseline-sharing groups it handles are automatically updated,
+// if there is one that is compatible with such item. Otherwise, a new baseline-sharing group is created,
+// compatible with the new item.
+class BaselineContext {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    BaselineContext(const RenderBox& child, ItemPosition preference, LayoutUnit ascent);
+    const BaselineGroup& sharedGroup(const RenderBox& child, ItemPosition preference) const;
+
+    // Updates the baseline-sharing group compatible with the item.
+    // We pass the item's baseline-preference to avoid dependencies with the LayoutGrid class, which is the one
+    // managing the alignment behavior of the Grid Items.
+    void updateSharedGroup(const RenderBox& child, ItemPosition preference, LayoutUnit ascent);
+
+private:
+    // Returns the baseline-sharing group compatible with an item.
+    // We pass the item's baseline-preference to avoid dependencies with the LayoutGrid class, which is the one
+    // managing the alignment behavior of the Grid Items.
+    // FIXME: Properly implement baseline-group compatibility.
+    // See https://github.com/w3c/csswg-drafts/issues/721
+    BaselineGroup& findCompatibleSharedGroup(const RenderBox& child, ItemPosition preference);
+
+    Vector<BaselineGroup> m_sharedGroups;
+};
+
+enum AllowedBaseLine { FirstLine, LastLine, BothLines };
+
+static inline bool isBaselinePosition(ItemPosition position)
```

Comment:
Nit: Consider putting these in an `inlines` file that includes the definition for ItemPosition::Baseline, etc.